sema4ai/vscode-client/src/robo/actionPackage.ts

```diff
@@ -384,3 +409,250 @@ async function getTemplate(actionServerLocation: string) {
+        return false;
+    }
+
+    return true;
+};
+
+export const publishActionPackage = async () => {
+    const workspaceDir = await askForWs();
+
+    await window.withProgress(
+        {
+            location: ProgressLocation.Notification,
+            title: "Publishing Action Package",
+            cancellable: false,
+        },
+        async (
+            progress: Progress<{ message?: string; increment?: number }>,
+            token: CancellationToken
+        ): Promise<void> => {
+            progress.report({ message: "Validating action server" });
+            const actionServerLocation = await downloadOrGetActionServerLocation();
+            if (!actionServerLocation) {
+                return;
+            }
+
+            progress.report({ message: "Validating authentication" });
+            const verifyLoginOutput = await verifyLogin(actionServerLocation);
+            if (!verifyLoginOutput || !verifyLoginOutput.logged_in) {
+                window.showErrorMessage(
+                    "Action Server Not authenticated to Control Room, run: Sema4ai: Authenticate the Action Server to Control Room"
+                );
+                return;
+            }
+
+            progress.report({ message: "Finding organizations" });
+            const organizations = await listOrganizations(actionServerLocation);
+            if (organizations.length === 0) {
+                return;
+            }
+
+            progress.report({ message: "Choose organization" });
+            const organization = await chooseOrganization(organizations);
+            if (!organization) {
+                return;
+            }
+
+            const tempDir = path.join(os.tmpdir(), "vscode-extension", Date.now().toString());
+            try {
+                fs.mkdirSync(tempDir, { recursive: true });
+
+                progress.report({ message: "Building package" });
+                const packagePath = await buildPackage(actionServerLocation, workspaceDir.uri.fsPath, tempDir);
+                if (!packagePath) {
+                    return;
+                }
+
+                progress.report({ message: "Uploading package to Control Room" });
+                const actionPackage = await uploadActionPackage(actionServerLocation, packagePath, organization.id);
+                if (!actionPackage) {
+                    return;
+                }
+
+                progress.report({ message: "Waiting for package to be verified" });
+                const verified = await waitUntilPackageVerified(
+                    actionServerLocation,
+                    actionPackage,
+                    organization.id,
+                    progress
+                );
+                if (!verified) {
+                    return;
+                }
+
+                progress.report({ message: "Getting changelog input" });
+                const changelogInput = await askUserForChangelogInput();
```

Comment:
Maybe we can start the `upload` without awaiting and then ask the user for the changelog input right away (and only then do the await on the `upload` / wait for verification (otherwise the user has to monitor the upload/verification to then provide that info, which I guess can be a bit annoying).

-- on the other hand, if we're going to evolve to a webview in a short time, maybe this would be optional at this point.

Reply:
I'll move the changelog input request to the start.